Add migration docs (#1099)
guardian/cdk: docs/migration-guide-ec2.md

---

The [GuEc2App](https://guardian.github.io/cdk/classes/index.GuEc2App.html)
pattern describes a web service with instances running in an auto-scaling group
(ASG), accessed via an Application Load Balancer (ALB).

It is an appropriate CDK target for many Guardian web services.

However, migrating running applications is not always straightforward. For
example, the app may use an ELB whereas the pattern uses an ALB.

For apps where avoiding downtime is critical, we recommend a _dual-stack
migration strategy_. The idea is to duplicate the key parts of the stack using
the pattern, test, and, once confident, switch DNS to point to the new stack.

**Stage 1: dual-stack**

```mermaid
graph TD
    A(example.gutools.co.uk)
    A --> B[Legacy ELB]
    A .-> C[GuEC2App ALB]
```

**Stage 2: switch DNS**

```mermaid
graph TD
    A(example.gutools.co.uk)
    A .-> B[Legacy ELB]
    A --> C[GuEC2App ALB]
```

**Stage 3: cleanup**

```mermaid
graph TD
    A(example.gutools.co.uk)
    A --> C[GuEC2App ALB]
```

A fully-worked out example can be found
[here](https://github.com/guardian/amigo/pulls?q=is%3Apr+is%3Aclosed+author%3Aakash1810+label%3Acfn-yaml-to-cdk).

> If you application includes a database, queues, S3 buckets, etc., you can either

Reviewer: ```suggestion
If your application includes a database, queues, S3 buckets, etc., you can either
```

    as part of the deployment.

6.  Test your app’s functionality via the new load balancer’s DNS name (this can
    be found as an output in the CFN console).

## Stage 2: switch DNS

_This stage ensures that you are able to manage your own DNS record(s) via CDK and updates your DNS record(s) to point at your new load balancer(s)._

_At the end of this stage, all traffic should be running via the new resources that we created in Stage 2. The equivalent old resources should still be part of your stack, but they should be redundant by the end of this stage._

_The instructions here assume that your DNS record is managed via NS1. To check if you are using NS1, you can use dig - dig +nssearch +noall +answer <TLD> where TLD is the top level domain. For example dig +nssearch +noall +answer gutools.co.uk._

1.  Start managing your DNS record via CDK by instantiating a GuCname construct

    Use dig to check the current properties of your DNS record, for example:

    ```
    $ dig +noall +answer CNAME security-hq.gutools.co.uk
    $ security-hq.gutools.co.uk. 3600	IN	CNAME	secur-loadb-re7mdt13o53h-879373669.eu-west-1.elb.amazonaws.com.
    ```

>     (3600 is the TTL, and `secur0loadb...` is the loadbalancer.)

Reviewer: ```suggestion
    (3600 is the TTL, and `secur-loadb...` is the loadbalancer.)
```

    Instantiate the GuCname class, ensuring that all properties match the
    properties in NS1. The CNAME should still point at your old load balancer’s
    DNS name for now.

    There's an example of this
    [here](https://github.com/guardian/security-hq/pull/336/files#diff-f38efbe4db4fb3d00ad6e3d6792a8d0e4fa818b3b974cb9d012ce2e81a242bfc).

    _N.B. if your template already includes a Guardian::DNS::RecordSet resource,
    you will need to:_

    - _Use the same [logical
      id](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html)
      when instantiating the GuCname via CloudFormation_
    - _Delete the Guardian::DNS::RecordSet resource in the same PR that the
      GuCname is added_

2.  Drop the TTL for your DNS record, to enable faster rollback if something goes wrong.

    Wait for the TTL to expire before proceeding. ([Example
    PR](https://github.com/guardian/security-hq/pull/338/files)).

3.  Update DNS entry for your so that it points to the new load balancer
    ([Example PR](https://github.com/guardian/security-hq/pull/339/files)).

4.  Run some tests to confirm that functionality still works as expected.

> 5.  Wait for a while before proceeding with Stage 6.

Reviewer: Minor, this confused me at first because I was looking for a `## Stage 6: xyz` section

```suggestion
5.  Wait for a while before proceeding with the next step.
```

    If something goes wrong at this point, we can still roll back by reverting
    the DNS change.

6.  Once you are confident that the new infrastructure is working correctly, increase the TTL for your DNS record again.

## Stage 3: cleanup

_This stage removes all redundant resources from your original template._

_At the end of this stage, the vast majority of your resources will be created
via the CDK pattern. Some resources, such as Dynamo tables, may still be defined
in your original CFN template._

1. View request count metrics for old vs new load balancers in CloudWatch.

   You should not proceed until you have confirmed that the old load balancer is
   receiving 0 requests.

2. Identify redundant resources in your original CFN template.

   For example, ASGs, LoadBalancers, previous alarms, etc.

3. Remove the redundant resources from your original CFN template.

   In simple cases it may be possible to remove the whole file. If so, you
   should also be able to remove the CfnInclude block and the associated
   dependency

4. Remove the gu:riffraff:new-asg tag from your new ASG.

5. Remove the `asgMigrationInProgress` parameter from your riff-raff.yaml.

> 6. Update your riff-raff.yaml to use the correct `amiParameter` (this will be AMI<Appname>)

Reviewer: The current version hides `<Appname>`!

```suggestion
6. Update your riff-raff.yaml to use the correct `amiParameter` (this will be `AMI<Appname>`)
```
